File: go/solver.go
package main

import (
	"bufio"
	"fmt"
	"strings"
	"sort"
	"io/ioutil"
)

var letters string = "abcdefg";
var words[] string;

func isValid(word string) bool {
	for _, char := range word {
		letter := string(char)
		if strings.Count(word, letter) > strings.Count(letters, letter) {
			return false
		}
	}

	return true
}

func loadValidWords() {
	bytes, _ := ioutil.ReadFile("dictionary.txt")
	content := string(bytes)
	scanner := bufio.NewScanner(strings.NewReader(content))

	for scanner.Scan() {
		if isValid(scanner.Text()) {
			words = append(words, scanner.Text())
		}
	}
}

var scores [26]int = [26]int {1, 3, 3, 2, 1, 4, 2, 4, 1, 8, 5, 1, 3, 1, 1, 3, 10, 1, 1, 1, 1, 4, 4, 8, 4, 10}

func getScore(word string) int {
	total := 0

	for _, char := range word  {
		total += scores[int(char) - 97]
	}
	
	return total
}

func sortWords() {
	sort.Slice(words, func(lower int, higher int) bool {
		return getScore(words[lower]) < getScore(words[higher])
	})
}

func displayWords() {
  for i := len(words); i > 0; i-- {
    fmt.Println(i, ". ", words[len(words) - i]);
	}
}
func main() {
  loadValidWords()
  sortWords()
  displayWords()
}

File: lua/solver.lua
local letters <const> = "abcdefg"

function count(word, target)
  return select(2, string.gsub(word, target, ""))
end

function valid(word) 
  for idx = 1, #word do
    local letter = word:sub(idx, idx)

    if count(word, letter) > count(letters, letter) then
      return false
    end
  end

  return true
end

function loadValidWords()
  local words = {}

  local file = assert(io.open("dictionary.txt", "r"))
  local content = file:read("*all")
  file:close()

  for word in content:gmatch("[^\n\r]+") do
    if valid(word) then
      words[#words + 1] = word
    end
  end

  return words
end

function getScore(word)
  local scores <const> = {1, 3, 3, 2, 1, 4, 2, 4, 1, 8, 5, 1, 3, 1, 1, 3, 10, 1, 
    1, 1, 11, 4, 4, 8, 4, 10}

  local total = 0

  for idx = 1, #word do
    total = total + scores[word:byte(idx) - 96]
  end

  return total
end

function sortWords(words)
  table.sort(words, function (a, b)
    return getScore(a) < getScore(b)
  end)
end

function displayWords(words)
  for idx = #words, 1, -1 do
    print(idx .. [[. ]] .. words[#words - idx + 1])
  end
end

function main()
  local words = loadValidWords()
  sortWords(words)
  displayWords(words)
end

main()
